device_tracker.ubus: use dhcp & iwinfo services instead of current multi-step process

rytilahti opened the pull request:
**Description:**
* Converted to use iwinfo & dhcp services, iwinfo allows instantaneous detection of changes in connected devices (coming directly from the wifi radio instead of using leases like before), making this tracker usable.
* This also allows more fine-grained permissions, so no more root user required (see updated documentation) for functionality.
* Better error handling and reporting. debug() reporting for important pieces for future development.
* Adapt to use setup_scanner() API.
Using dhcp service allows getting the wanted information without
reading extra files and giving extra permissions.

**Related issue (if applicable):** fixes #2175, #4495, #1492

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1819

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

rytilahti committed:
WIP: device_tracker.ubus: use dhcp service instead of current multi-step process

Using dhcp service allows getting the wanted information without
reading extra files and giving extra permissions.

This commit also adds some more error checking & debug facilities for further development.

TBD: As MAC is not properly encoded, the vendor value can't be fetched currently.

andriej commented:
Is DHCP-lease good option anyway? 
Lease can be active while host may already be far away from router…

I have some openwrt-side script which I may share some time after clean-up.
It does check if known MAC is still receiving packets or not (ping-independent).

rytilahti commented:
No it is not good but that's what the original implementation was doing. I was adapting it to be more constrained until I noticed that it's unusable for device tracking :-)

I just wanted to put this up as RFC if someone has a better idea how to do that, as I couldn't find any sane way to make it happen.

kellerza commented:
One possibility would be to change the arp timeout, it was discussed [here](https://github.com/home-assistant/home-assistant/pull/4193#issuecomment-258850487) and was set to 4  minutes in the end.

If not possible to set this on your device, can we close this PR?

rytilahti commented:
I bet it's somehow controllable but no idea how easily it's done. Anycase, this PR can still be useful for those people using this tracker, as it greatly simplifies the code while still keeping the functionality.

rytilahti committed:
ubus: rework to use setup_scanner api, and dhcp & iwinfo services

This commit makes ubus tracker to use dhcp and iwinfo
services for device tracking, allowing instantaneous device recognition,
thanks to using wifi radio for tracking the connected devices.
dhcp service is required for hostname fetching.

Ports the component also to use setup_scanner() api to provide
more information about the tracked devices.

rytilahti committed:
lint fixes

rytilahti commented:
I modified the tracker to use iwinfo interface for accessing tracked devices. This allows instantaneous device tracking as the information is coming directly from the radio(s).

rytilahti committed:
Loop over clients to allow clients without leases, may happen with some devices that they do not renew their leases enough often..

rytilahti committed:
ubus: do not do i/o on setup_scanner(), try to re-login on expired/failed sessions

rytilahti committed:
use mac as name, as it gets saved in the known_devices.yaml

rytilahti committed:
Fix lint warnings

rytilahti committed:
ubus: change warnings & infos to debugs, document login

rytilahti committed:
ubus: do not bail out on ConnectionError

kellerza commented:
@rytilahti this device_tracker has got so much protocol specific features in it at the moment, you will probably have to start moving it to an external library

@balloob's final call

balloob reviewed (changes requested):
I agree with @kellerza, we should extract the protocol logic into a 3rd party lib.

rytilahti commented:
Which parts you propose to be outed into a separate lib? Parts of the logic, or just the ubus communication? That is, would it be enough to create a python-ubus library for communication and use that inside the component?

rytilahti committed:
ubus: use request ids & cleanup the code a bit, preparation for outing comms to a separate lib, fixes also a crash when unable to fetch interfaces

homeassistant commented:
Hi @rytilahti, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5277).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @rytilahti, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5277).

Once you do that we will be able to review and accept this pull request.

Thanks!

balloob commented:
> That is, would it be enough to create a python-ubus library for communication and use that inside the component?

Yes. Move things like UbusStatus, UbusException, `_get_devices`, `_get_session_id` etc.

In your own lib you also don't have to pass Hound ;-)

andrey-git commented:
@rytilahti any updates on this PR?

rytilahti commented:
Unfortunately no, have not yet had time to create a library for this.

andrey-git commented:
@rytilahti do you plan to continue working on this or should the PR be closed?

rytilahti commented:
@andrey-git I'm planning to work on this but I'm currently occupied with other things so it may take a while before I can create a proper ubus lib and port this on top of it.

balloob commented:
Alright, we'll close this and you can open a new one when ready to finish it.
